pkg/controller/persistentpodstate/persistent_pod_state_controller.go

```diff
@@ -0,0 +1,451 @@
+	}
+	nodeTopologyKeys := sets.NewString()
+	// required node labels
+	if persistentPodState.Spec.RequiredPersistentTopology != nil {
+		nodeTopologyKeys.Insert(persistentPodState.Spec.RequiredPersistentTopology.NodeTopologyKeys...)
+	}
+	// preferred node labels
+	for _, item := range persistentPodState.Spec.PreferredPersistentTopology {
+		nodeTopologyKeys.Insert(item.Preference.NodeTopologyKeys...)
+	}
+	// create sts scenario
+	for _, pod := range pods {
+		// 1. pod not ready, continue
+		if !podutil.IsPodReady(pod) || pod.Spec.NodeName == "" {
+			continue
+		}
+		// 2. check old pod state
+		if podState, ok := newStatus.PodStates[pod.Name]; ok {
+			currentKeys := sets.NewString()
+			for key := range podState.NodeTopologyLabels {
+				currentKeys.Insert(key)
+			}
+			// already recorded, no need to regenerate
+			if podState.NodeName == pod.Spec.NodeName && nodeTopologyKeys.Equal(currentKeys) {
+				continue
+			}
+		}
+		// 3. create new pod state
+		newState, err := r.createPodState(pod, nodeTopologyKeys)
+		if err != nil {
+			continue
+		}
+		// 4. store PodState
+		newStatus.PodStates[pod.Name] = newState
+	}
+
+	// scale down statefulSet scenario
+	if persistentPodState.Spec.PersistentPodStateRetentionPolicy != appsv1alpha1.PersistentPodStateRetentionPolicyWhenDeleted {
+		for podName := range newStatus.PodStates {
+			index, err := parseStsPodIndex(podName)
+			if err != nil {
+				klog.Errorf("parse PersistentPodState(%s/%s) podName(%s) failed: %s",
+					persistentPodState.Namespace, persistentPodState.Name, podName, err.Error())
+				continue
+			}
+			if isInStatefulSetReplicas(index, innerSts) {
+				continue
+			}
+			// others will be deleted for scaling down sts
+			// if pod not exists, then delete PersistentPodState
+			if _, ok := pods[podName]; !ok {
+				delete(newStatus.PodStates, podName)
+			}
+		}
+	}
+	return ctrl.Result{}, r.updatePersistentPodStateStatus(persistentPodState, *newStatus)
+}
+
+func (r *ReconcilePersistentPodState) updatePersistentPodStateStatus(pps *appsv1alpha1.PersistentPodState, newStatus appsv1alpha1.PersistentPodStateStatus) error {
+	if reflect.DeepEqual(pps.Status, newStatus) {
+		return nil
+	}
+	// update PersistentPodState status
+	persistentPodStateClone := &appsv1alpha1.PersistentPodState{}
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := r.Client.Get(context.TODO(), client.ObjectKey{Namespace: pps.Namespace, Name: pps.Name}, persistentPodStateClone); err != nil {
+			return err
+		}
+		persistentPodStateClone.Status = newStatus
+		return r.Client.Status().Update(context.TODO(), persistentPodStateClone)
+	}); err != nil {
+		klog.Errorf("update PersistentPodState(%s/%s) status failed: %s", pps.Namespace, pps.Name, err.Error())
+		return err
+	}
+	klog.V(3).Infof("update PersistentPodState(%s/%s) status from(%s) -> to(%s) success", pps.Namespace, pps.Name, util.DumpJSON(pps.Status), util.DumpJSON(newStatus))
```

Comment:
Should we print the status of all pods topologies here? It may at lease be `V(5)` and how about print the change states if we can?